test/Test/Pos/Types/CoinSpec.hs

```diff
@@ -0,0 +1,190 @@
+
+coinAdditionWorks :: C.SafeCoinPairSum -> Property
+coinAdditionWorks =
+    let longFunction =
+            C.mkCoin .
+            uncurry (+) .
+            bimap C.unsafeGetCoin C.unsafeGetCoin .
+            C.getPairSum
+    in longFunction .=. (uncurry C.unsafeAddCoin . C.getPairSum)
+
+underflowInSubCausesError :: C.CoinPairOverflowSub -> Expectation
+underflowInSubCausesError =
+    shouldThrowException (uncurry C.unsafeSubCoin . C.get2CSub) fatalException
+
+coinSubtractionWorks :: C.SafeCoinPairSub -> Property
+coinSubtractionWorks =
+    let longFunction =
+            C.mkCoin .
+            uncurry (-) .
+            bimap C.unsafeGetCoin C.unsafeGetCoin .
+            C.getPairSub
+    in longFunction .=. (uncurry C.unsafeSubCoin . C.getPairSub)
+
+overflowInMulCausesError :: C.CoinPairOverflowMul -> Expectation
+overflowInMulCausesError =
+    shouldThrowException (uncurry C.unsafeMulCoin . C.get2CMul) fatalException
+
+coinProductWorks :: C.SafeCoinPairMul -> Property
+coinProductWorks =
+    let longFunction =
+            C.unsafeIntegerToCoin .
+            uncurry (*) .
+            bimap C.coinToInteger identity .
+            C.getPairMul
+    in longFunction .=. (uncurry C.unsafeMulCoin . C.getPairMul)
+
+overflowIntegerCausesError :: C.IntegerToCoinOverflow -> Expectation
+overflowIntegerCausesError =
+    shouldThrowException (C.unsafeIntegerToCoin . C.getLargeInteger) fatalException
+
+coinToIntegralToCoin :: C.Coin -> Property
+coinToIntegralToCoin = C.mkCoin . C.unsafeGetCoin .=. identity
+
+wordToCoinToWord :: Word64 -> Property
+wordToCoinToWord = C.unsafeGetCoin . C.mkCoin .=. identity
+
+coinToIntegerToCoin :: C.Coin -> Property
+coinToIntegerToCoin = C.unsafeIntegerToCoin . C.coinToInteger .=. identity
+
+integerToCoinToInteger :: C.IntegerToCoinNoOverflow -> Property
+integerToCoinToInteger =
+    C.Integer . C.coinToInteger . C.unsafeIntegerToCoin . C.getInteger .=. identity
+
+sumCoinsIsNeverNegative :: [C.Coin] -> Property
+sumCoinsIsNeverNegative = (>= 0) . C.sumCoins .=. const True
+
+------------------------------------------------------------------------------------------
+-- CoinPortion
+------------------------------------------------------------------------------------------
+
+overOrUnderflowDoubleCausesError :: C.LessThanZeroOrMoreThanOne -> Expectation
+overOrUnderflowDoubleCausesError =
+    shouldThrowException (C.unsafeCoinPortionFromDouble . C.getDouble) fatalException
+
+coinPortionToDoubleToPortion :: C.CoinPortion -> Property
+coinPortionToDoubleToPortion =
+    C.unsafeCoinPortionFromDouble . C.coinPortionToDouble .=. identity
+
+-- | Combinator to facilitate the 'wordToPortion' property.
+(=<<<) :: Monad m => (a -> b) -> (c -> m a) -> (c -> m b)
+(=<<<) f g = \c -> do
+    a <- g c
+    return . f $ a
+
+infixr 5 =<<<
+
+wordToPortionToWord :: C.SafeWord -> Property
+wordToPortionToWord =
+    (C.SafeWord . C.getCoinPortion =<<< C.mkCoinPortion . C.getSafeWord) >=. (pure @Maybe)
```

Comment:
Come on, don't mock me just yet.
I wrote it like this because I only had `pure` before, and I had to think about what monad gave me a free `Eq` instance. I realize this isn't ideal as `Maybe` is one of the most trivial monads, but the idea here was to test that conversions work fine, and the monad used to simulate an environmentdoesn't matter.